File: PC/Effect.cs
using System.Linq;
using UnityEngine;

namespace Weather
{
    public class Effect
    {
        public EffectDescriptor Desc { get; private set; }
        private GameObject GameObject { get; set; }
        public bool Enabled { get; set; }

        public bool ShowInMenu { get; set; } = true;
        public bool ShowInGame { get; set; } = true;

        public Effect(EffectDescriptor effectDescriptor, GameObject gameObject, bool enabled)
        {
            Desc = effectDescriptor;
            GameObject = gameObject;
            Enabled = enabled;
            
        }
        public void SetActiveRefs(bool force = false)
        {
            Plugin.Log.Info("Setting Active Refs " + Desc.effectName + " " + WeatherSceneInfo.CurrentScene.name);
            Enabled = EffectModel.GetEffectEnabledByName(Desc.effectName) || force;
            if(!Enabled)
            {
                GameObject.SetActive(false);
                return;
            }         
            
            switch (WeatherSceneInfo.CurrentScene.name)
            {
                case Plugin.Menu:
                    Plugin.Log.Info(Desc.effectName + (Desc.worksInMenu && ShowInMenu));
                    GameObject.SetActive((Desc.worksInMenu && ShowInMenu));
                    break;
                case Plugin.Game:
                    Plugin.Log.Info(Desc.effectName + (Desc.worksInGame && ShowInGame));
                    GameObject.SetActive((Desc.worksInGame && ShowInGame));
                    break;
            }
        }
        public void SetSceneMaterials()
        {
            var grab = GameObject.transform.GetChild(0).Find("NotesShader");
            if (grab == null) return;
            var mrs = Resources.FindObjectsOfTypeAll<MeshRenderer>();
            foreach (var mr in mrs)
            {
                if (mr.material.name.Contains("Note") || mr.gameObject.name.Contains("building") || mr.gameObject.name.Contains("speaker"))
                {
                    TrySetNoteMaterial(mr);
                }
            }
        }
        public void RemoveSceneMaterials()
        {
            var grab = GameObject.transform.GetChild(0).Find("NotesShader");
            if (grab == null) return;
            var mrs = Resources.FindObjectsOfTypeAll<MeshRenderer>();
            foreach (var mr in mrs)
            {
                if (!mr.material.name.Contains("Note") && !mr.gameObject.name.Contains("building") &&
                    !mr.gameObject.name.Contains("speaker")) continue;
                var mats = mr.sharedMaterials.ToList();
                mats.RemoveAt(1);
                mr.materials = mats.ToArray();
            }
        }
        public void TrySetNoteMaterial(MeshRenderer mr)
        {
            if (!Enabled) return;
            var grab = GameObject.transform.GetChild(0).Find("NotesShader");
            if (grab == null) return;
            var noteMaterial = grab.GetComponent<MeshRenderer>().material;
            var mats = mr.sharedMaterials.ToList();
            if(mats.Count == 1)
                mats.Add(noteMaterial);
            mr.materials = mats.ToArray();
        }
    }
}

File: Quest/src/Effect.cpp
#include "Effect.hpp"
#include "EffectDescriptor.hpp"
#include "WeatherSceneInfo.hpp"
#include "EffectModel.hpp"
#include "include/Logger.hpp"
#include "UnityEngine/Material.hpp"
#include "UnityEngine/Transform.hpp"
#include "System/Collections/Generic/List_1.hpp"
#include "System/Linq/Enumerable.hpp"
#include "Config.hpp"

#include "beatsaber-hook/shared/utils/il2cpp-utils.hpp"
using namespace UnityEngine;
using namespace System::Collections::Generic;
using namespace System::Linq;

DEFINE_CLASS(Weather::Effect);

void Weather::Effect::ctor(Weather::EffectDescriptor* _effectDescriptor, GameObject* _gameObject, bool _enabled) {
    this->enabled = _enabled;
    this->Desc = _effectDescriptor;
    this->gameObject = _gameObject;
}

void Weather::Effect::SetActiveRefs(bool force) {
    Weather::GenericLogger::Log("SetActiveRefs");
    this->Grab = gameObject->get_transform()->GetChild(0)->Find(il2cpp_utils::createcsstr("NotesShader"));
    enabled = EffectModel::GetEffectEnabledByName(Desc->EffectName);
    if(force) enabled = force;
    if(!enabled)
    {
        Weather::GenericLogger::Log("NotFound");
        gameObject->SetActive(false);
        return;
    }
    if(to_utf8(csstrtostr(WeatherSceneInfo::CurrentScene.get_name())) == "MenuViewControllers")
    {
        gameObject->SetActive((Desc->WorksInMenu && showInMenu));
    }
    if(to_utf8(csstrtostr(WeatherSceneInfo::CurrentScene.get_name())) == "GameCore")
    {
        gameObject->SetActive((Desc->WorksInGame && showInGame));
    }
}
//Unfilled since Ingame UI isnt implemented yet
void Weather::Effect::SetSceneMaterials()
{

}

void Weather::Effect::RemoveSceneMaterials()
{
    
}
void Weather::Effect::TrySetNoteMateral(MeshRenderer* mr) {
    if(!Grab || !Grab->GetComponent<MeshRenderer*>()) return;
    if(!getConfig().config["enableShaderAdditions"].GetBool()) return;
    if(!enabled) return;
    Material* notemat = Grab->GetComponent<MeshRenderer*>()->get_material();
    Array<Material*>* mats = mr->get_materials();
    Array<Material*>* newMats = mats->NewLength(2);
    newMats->values[0] = mats->values[0];
    newMats->values[1] = notemat;
    mr->set_materials(newMats);
}